### PR TITLE
ENH: Add method to write GeoDataFrame to PostGIS

### PR DESCRIPTION
**Pull request by mazzma12**
Hi, 

This is my first attempt to contribute to this project. The goal is to seamlessly ingest a `GeoDataFrame` object into a PostGIS database in a similar way that `pandas` interacts with PostgreSQL. 
I inspired from #457  and #761 to come up with this method that 

1. Uses `geoalchemy2` as a requirement (only added to `requirements.test.txt`) 
2. Make a copy of the original gdf
3. Infer the PostGiS geom type from `geopandas.GeoDataFrame.geom_type()` 
4. Convert the CRS to a PostGIS SRID format
5. Pass arguments to all arguments to `pandas.to_sql()` 

### Notes
* Unlike #457, this does not attempt to parse converts shapely objects from another column that the geometry.  Geopandas is designed to work off-the shell with only one geometry column, is then up to the user to change the types. 
* It does a full copy of the frame in the first place because the geometry has to be converted and one shall not mutate the object. This can be heavy in memory for large frames with large shapes. One could add a copy argument default to `True` or maybe you have another 

Ofc this is a WIP and If you like it I can continue documenting it and writing some tests. 

Cheers!

 (Ping @jdmcbr )

PS : I found that the way tests are handled is complicated as it requires to have a PostGIS instance running on the same machine. I came up with a `docker-compose` setup of a PostGIS image to emulate that - look at the `ci` branch.

### CODE REVIEW

File: geopandas/tools/util.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import fiona
```
Comment (mazzma12) [inline geopandas/tools/util.py]: I guess this is the import that makes the build fails

**Comment by mazzma12** [comment]
@jdmcbr any advice on this one? I couldn't find a documentation on how to build tests for SQL databases. I wasn't successful either at reproducing them in my environment, so I might need some help on this one. 

Also, I do not understand how to restart codecov tests. To me the documentation coverage is large enough, I do not get why it fails. Perhaps the drop is due to the notebook, I could remove it.

File: geopandas/tools/util.py [hunk 2]
```diff
@@ -42,3 +44,25 @@ def collect(x, multi=False):
...
+    srid : int, -1  if unsuccessful for compatibility with PostGIS
+    """
+    srid = -1
+
+    if isinstance(crs, str):
+        crs = from_string(crs)
+    if isinstance(crs, dict):
+        srid = int(crs.get('init').split(':')[-1])
+
+    return srid
```
Comment (jdmcbr) [inline geopandas/tools/util.py]: There's now a `tools.crs` module where this would fit well.

File: requirements.test.txt [hunk 3]
```diff
@@ -9,3 +9,6 @@ pytest-cov
 codecov
 rtree>=0.8
 pysal
+geoalchemy2
+boto3
+fiona==1.7.12 # Fixing version to avoid conda installing alpha #853
```
Comment (jdmcbr) [inline requirements.test.txt]: I don't follow the explanatory comment, and would prefer not to be pinning a specific version here if not absolutely necessary.

File: geopandas/io/sql.py [hunk 4]
```diff
@@ -63,3 +65,111 @@ def load_geom(x):
...
+    ----------
+    .. [1] http://docs.sqlalchemy.org
+    .. [2] https://www.python.org/dev/peps/pep-0249/
+    .. [3] https://geoalchemy-2.readthedocs.io/en/latest/
+
+    Notes
+    ------
+    Geometry is converted to `WKTElements` object using `geoalchemy2` library.
+    The original frame is copied to not mutate it.
+    Only one column with shapely objets is supported.
```
Comment (jdmcbr) [inline geopandas/io/sql.py]: "objets" -> "objects"

File: geopandas/io/sql.py [hunk 5]
```diff
@@ -63,3 +65,111 @@ def load_geom(x):
...
+    pandas.read_sql : pass kwargs to this function
+
+    References
+    ----------
+    .. [1] http://docs.sqlalchemy.org
+    .. [2] https://www.python.org/dev/peps/pep-0249/
+    .. [3] https://geoalchemy-2.readthedocs.io/en/latest/
+
+    Notes
+    ------
```
Comment (jdmcbr) [inline geopandas/io/sql.py]: For nice auto-doc building, please align the number of `-` with the number of characters in the section header.

File: geopandas/io/sql.py [hunk 6]
```diff
@@ -63,3 +65,111 @@ def load_geom(x):
...
+    Geometry is converted to `WKTElements` object using `geoalchemy2` library.
+    The original frame is copied to not mutate it.
+    Only one column with shapely objets is supported.
+
+    Examples:
+    ---------
```
Comment (jdmcbr) [inline geopandas/io/sql.py]: Also for nice auto-doc, please remove the `:` here.

File: requirements.test.txt (hunk 3 above)
Comment (mazzma12) [reply requirements.test.txt]: I totally agree there, there just was a build error at the time due to fiona I think it should be solved by now

**Comment by mazzma12** [comment]
Hi @jdmcbr sorry for late reply, I agree with you review and I updated the PR :
* Fix docstring typos
* unpin fiona version
* move utils
* add your tests

Regarding the test with PostGIS I am unable to run them in local to check if they pass. I use docker for a local postgis instance (see the `test_docker` branch for the config), but the config for tests is hardocded so I can't change the host -and got skipped tests.

**Comment by mazzma12** [comment]
Actually I think the test fails because the input df nybb does not have a crs value expected : 
{'datum': 'NAD83',
 'lat_0': 40.16666666666666,
 'lat_1': 40.66666666666666,
 'lat_2': 41.03333333333333,
 'lon_0': -74,
 'no_defs': True,
 'proj': 'lcc',
 'units': 'us-ft',
 'x_0': 300000,
 'y_0': 0}

File: geopandas/io/tests/test_io.py [hunk 7]
```diff
@@ -213,3 +213,58 @@ def test_empty_shapefile(self, tmpdir):
...
+        try:
+            db_name = "nybb4326_write"
+            df_reproj.to_postgis(db_name, con, if_exists="replace")
+            back = read_postgis("SELECT * FROM {};".format(db_name), con,
+                                geom_col="geometry")
+        finally:
+            con.close()
+
+        validate_boro_df(back)
+        assert back.crs == crs
```
Comment (jdmcbr) [inline geopandas/io/tests/test_io.py]: This produces the following on my computer:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

self = <geopandas.io.tests.test_io.TestIO object at 0x7f9591bf1710>

    def test_to_postgis_set_srid(self):
    
        con = connect('test_geopandas')
        if con is None or not create_postgis(self.df):
            raise pytest.skip()
    
        crs = {"init": "epsg:4326"}
        df_reproj = self.df.to_crs(crs)
    
        try:
            db_name = "nybb4326_write"
>           df_reproj.to_postgis(db_name, con, if_exists="replace")

test_io.py:244: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../geodataframe.py:343: in to_postgis
    return geopandas.io.sql.write_postgis(self, name, con, **kwargs)
../sql.py:155: in write_postgis
    temp_df.to_sql(name, con, **kwargs)
/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:2130: in to_sql
    dtype=dtype)
/usr/local/lib/python3.6/dist-packages/pandas/io/sql.py:450: in to_sql
    chunksize=chunksize, dtype=dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.io.sql.SQLiteDatabase object at 0x7f9591bf1b38>
frame =    BoroCode                        ...                                                                   geometry
0   ...                   ...                          MULTIPOLYGON (((-73.89680883223774 40.79580844...

[5 rows x 5 columns]
name = 'nybb4326_write', if_exists = 'replace', index = True, index_label = None
schema = None, chunksize = None
dtype = {'geometry': Geometry(geometry_type='MULTIPOLYGON', srid=4326)}

    def to_sql(self, frame, name, if_exists='fail', index=True,
               index_label=None, schema=None, chunksize=None, dtype=None):
        """
            Write records stored in a DataFrame to a SQL database.
    
            Parameters
            ----------
            frame: DataFrame
            name: string
                Name of SQL table.
            if_exists: {'fail', 'replace', 'append'}, default 'fail'
                fail: If table exists, do nothing.
                replace: If table exists, drop it, recreate it, and insert data.
                append: If table exists, insert data. Create if it does not exist.
            index : boolean, default True
                Write DataFrame index as a column
            index_label : string or sequence, default None
                Column label for index column(s). If None is given (default) and
                `index` is True, then the index names are used.
                A sequence should be given if the DataFrame uses MultiIndex.
            schema : string, default None
                Ignored parameter included for compatibility with SQLAlchemy
                version of ``to_sql``.
            chunksize : int, default None
                If not None, then rows will be written in batches of this
                size at a time. If None, all rows will be written at once.
            dtype : single type or dict of column name to SQL type, default None
                Optional specifying the datatype for columns. The SQL type should
                be a string. If all columns are of the same type, one single value
                can be used.
    
            """
        if dtype and not is_dict_like(dtype):
            dtype = {col_name: dtype for col_name in frame}
    
        if dtype is not None:
            for col, my_type in dtype.items():
                if not isinstance(my_type, str):
                    raise ValueError('%s (%s) not a string' % (
>                       col, str(my_type)))
E                   ValueError: geometry (geometry(MULTIPOLYGON,4326)) not a string
```

File: geopandas/io/tests/test_io.py [hunk 8]
```diff
@@ -213,3 +213,58 @@ def test_empty_shapefile(self, tmpdir):
...
+
+        try:
+            db_name = "nybb_nocrs_write"
+            df_noproj.to_postgis(db_name, con, if_exists="replace")
+            back = read_postgis("SELECT * FROM {};".format(db_name), con,
+                                geom_col="geometry")
+        finally:
+            con.close()
+        validate_boro_df(back)
+        assert back.crs is None
```
Comment (jdmcbr) [inline geopandas/io/tests/test_io.py]: And:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

self = <geopandas.io.tests.test_io.TestIO object at 0x7f9592399e10>

    def test_to_postgis_no_srid(self):
    
        con = connect('test_geopandas')
        if con is None or not create_postgis(self.df):
            raise pytest.skip()
    
        df_noproj = self.df.copy()
        df_noproj.crs = None
    
        try:
            db_name = "nybb_nocrs_write"
>           df_noproj.to_postgis(db_name, con, if_exists="replace")

test_io.py:264: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../geodataframe.py:343: in to_postgis
    return geopandas.io.sql.write_postgis(self, name, con, **kwargs)
../sql.py:155: in write_postgis
    temp_df.to_sql(name, con, **kwargs)
/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:2130: in to_sql
    dtype=dtype)
/usr/local/lib/python3.6/dist-packages/pandas/io/sql.py:450: in to_sql
    chunksize=chunksize, dtype=dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.io.sql.SQLiteDatabase object at 0x7f95924b9940>
frame =    BoroCode                        ...                                                                   geometry
0   ...                   ...                          MULTIPOLYGON (((1012821.805786133 229228.26458...

[5 rows x 5 columns]
name = 'nybb_nocrs_write', if_exists = 'replace', index = True
index_label = None, schema = None, chunksize = None
dtype = {'geometry': Geometry(geometry_type='MULTIPOLYGON')}

    def to_sql(self, frame, name, if_exists='fail', index=True,
               index_label=None, schema=None, chunksize=None, dtype=None):
        """
            Write records stored in a DataFrame to a SQL database.
    
            Parameters
            ----------
            frame: DataFrame
            name: string
                Name of SQL table.
            if_exists: {'fail', 'replace', 'append'}, default 'fail'
                fail: If table exists, do nothing.
                replace: If table exists, drop it, recreate it, and insert data.
                append: If table exists, insert data. Create if it does not exist.
            index : boolean, default True
                Write DataFrame index as a column
            index_label : string or sequence, default None
                Column label for index column(s). If None is given (default) and
                `index` is True, then the index names are used.
                A sequence should be given if the DataFrame uses MultiIndex.
            schema : string, default None
                Ignored parameter included for compatibility with SQLAlchemy
                version of ``to_sql``.
            chunksize : int, default None
                If not None, then rows will be written in batches of this
                size at a time. If None, all rows will be written at once.
            dtype : single type or dict of column name to SQL type, default None
                Optional specifying the datatype for columns. The SQL type should
                be a string. If all columns are of the same type, one single value
                can be used.
    
            """
        if dtype and not is_dict_like(dtype):
            dtype = {col_name: dtype for col_name in frame}
    
        if dtype is not None:
            for col, my_type in dtype.items():
                if not isinstance(my_type, str):
                    raise ValueError('%s (%s) not a string' % (
>                       col, str(my_type)))
E                   ValueError: geometry (geometry(MULTIPOLYGON,-1)) not a string

/usr/local/lib/python3.6/dist-packages/pandas/io/sql.py:1475: ValueError
```

**Comment by mazzma12** [comment]
This version conflict is really odd .... I can try to parse the dtype first but I have no idea how to convert the geolachmey type engine to string. Gotta dive into the code

**Comment by mazzma12** [comment]
I pulled and merged master to get your last changes. I also struggle to investigate in this change of dtype parsing. You were right pandas version 0.23.4 enforce `str` types in dtype values, whereas `SQLAlchemy types` were allowed in `0.23.0`. 
Do you have some documentation about this change ? I could not find any in the release notes. I tried to git blame, but lots of change occures since then and could only find this PR https://github.com/pandas-dev/pandas/pull/13614/files

**Comment by leblancfg** [comment]
Just a little report, and a ping. I've been using this branch for a week now, great job @mazzma12 ! About a dozen tables written to PostGIS so far, and no errors &mdash; at least on my end, it seems at least the "happy path" is well covered.

Hard to understand where the development in the branch is right now (or is it in your local `:ci` branch, @mazzma12 ?), but do let me know if you need a hand somewhere.

**Comment by mazzma12** [comment]
> Just a little report, and a ping. I've been using this branch for a week now, great job @mazzma12 ! About a dozen tables written to PostGIS so far, and no errors — at least on my end, it seems at least the "happy path" is well covered.

Happy to see it's working on your side. My PostGIS use-cases are usually limited to Point and Polygon, did you manage to try other features? 

> Hard to understand where the development in the branch is right now (or is it in your local `:ci` branch, @mazzma12 ?), but do let me know if you need a hand somewhere.

Currently the development is staging, mainly because of the tests. I only run PostGIS in docker and it was not possible until recently to change PostGIS config from localhost (silved by #906)

The limiting factor now is that the test I have re implemented covers only the usecases I have met in GIS area, which my not be the defacto standard. For example the dataframe "boro" is used in geopandas testing package, but its crs does not conform the standard I have met in other packages and the tests are failing. (standard used at pyproj I think). 

Also, there was a change in pandas on how the `dtypes` argument is parsed. It only accepts `str` now which was different from what I have implemented at the time. 

If you are willing to help on this it will probably unblock the situation :)

**Comment by martinfleis** [comment]
Closed by #1248. Thank you all.

**Pull request closed without merging**